File: Compiler/src/errors/SyntaxError.ts
import { Logger } from '../utils/Logger';
import { ParserPointer } from '../utils/ParserPointer';
import { Pointer } from '../utils/Pointer';

type ErrorLog = {
	filename: string;
	lines: ErrorLine[];
};

type ErrorLine = {
	line: number;
	content: string;
};

type ErrorData = {
	lineError: number;
	reason: string;
};

export class SyntaxError extends Logger {
	constructor(private pointer: ParserPointer, private data: ErrorData) {
		super();

		this.setup();
	}

	private getLines(start: number) {
		const lines: ErrorLine[] = [];

		if (start > 0) {
			for (let i of [1, 2, 3, 4, 5].reverse()) {
				const lineNumber = start - i;
				if (lineNumber < 0) continue;

				const line = this.pointer.getLine(lineNumber);
				if (line === undefined) continue;

				lines.push({
					line: lineNumber,
					content: line,
				});
			}
		}

		lines.push({
			line: start,
			content: this.pointer.getLine(start),
		});

		for (let i of [1, 2, 3, 4, 5]) {
			const lineNumber = start + i;

			const line = this.pointer.getLine(lineNumber);
			if (line === undefined) continue;

			lines.push({
				line: lineNumber,
				content: line,
			});
		}

		return lines;
	}

	private logError(data: ErrorLog) {
		const { filename } = data;
		const { block } = this;

		const { lineError } = this.data;

		const currentLine = `in ${this.warn(`line ${lineError}`)}`;

		console.log(
			block(this.error('Error')),
			this.ctx(`SyntaxError on ${this.error(filename)} ${currentLine}`)
		);

		for (let { line, content } of data.lines) {
			const lineWarn = this.warn('Line ' + line);

			console.log(block(lineWarn), content);
		}

		console.log(block(this.info('Info')), this.ctx(this.data.reason));
	}

	private setup() {
		this.pointer = this.pointer as ParserPointer;

		const { pointer } = this;
		const { lineError } = this.data;

		this.logError({
			filename: pointer.filename,
			lines: this.getLines(lineError),
		});

		process.exit(1);
	}
}

File: Compiler/src/errors/syntax_error.rs
use std::process;

use serde_derive::Serialize;
use serde_json::to_string;

#[derive(Debug, Serialize)]
pub struct ErrorLine {
    pub line: usize,
    pub lineContent: String,
}

#[derive(Serialize)]
pub struct SyntaxError {
    r#type: String,
    reason: String,
    filename: String,
    lineNumber: usize,
    lineError: String,
    lines: Vec<ErrorLine>,
}

impl SyntaxError {
    pub fn new(filename: String, lines: Vec<Vec<String>>, line: usize, reason: String) {
        let err = SyntaxError {
            r#type: "SyntaxError".to_string(),
            reason,
            filename,
            lineNumber: line,
            lineError: SyntaxError::getLine(lines.clone(), (line - 1) as usize),
            lines: SyntaxError::getLines(lines.clone(), line as i64 - 1),
        };

        println!("{}", to_string(&err).unwrap());

        process::exit(1);
    }

    fn getLine(lines: Vec<Vec<String>>, position: usize) -> String {
        let currentLine = lines.get(position);

        if let Some(content) = currentLine {
            return content.join("");
        }

        "".to_string()
    }

    fn getLines(lines: Vec<Vec<String>>, line: i64) -> Vec<ErrorLine> {
        let mut errorLines: Vec<ErrorLine> = Vec::new();

        if line - 1 > 0 {
            for i in [5, 4, 3, 2, 1] {
                if line - 1 - i > 0 {
                    errorLines.push(ErrorLine {
                        line: (line - i) as usize,
                        lineContent: SyntaxError::getLine(lines.clone(), (line - 1 - i) as usize),
                    })
                }
            }
        }

        errorLines.push(ErrorLine {
            line: line as usize,
            lineContent: SyntaxError::getLine(lines.clone(), (line - 1) as usize),
        });

        for i in [1, 2, 3, 4, 5] {
            let lineContent = SyntaxError::getLine(lines.clone(), (line - 1 + i) as usize);

            if lineContent != "" {
                errorLines.push(ErrorLine {
                    line: (line + i) as usize,
                    lineContent,
                });
            }
        }

        return errorLines;
    }
}
